### ISSUE TITLE
evm symbolic memory research

### ISSUE THREAD

**Issue by mossberg** [trailofbits/manticore#736]

**Comment by feliam** [comment]
https://github.com/trailofbits/manticore/pull/771
Using smtlib arrays for both memory(array 8->256) and storage(array 256->256) allow us to pass the work to the solver. So you can now 'access' memory with unconstrained indexes hoping they will become constrained after the access.

Previously if all the input was symbolically free you could not get the smtlib description an EVM chunk of assembler. We used to try solving indexes early.

**Comment by mossberg** [comment]
by "solving indexes early" are you talking about how manticore would do a sampled concretization of the various memory locations that _could_ be accessed?

**Comment by feliam** [comment]
Yes. Linux memory implementation would solve _all_ (normally not just a sample) possible targets of the symbolic index on a memory access and build the resultant expression. That's eagerly.
The lazy option is to express the same thing in terms of an smtlib array. 

For both cases the problem is when the symbolic index has too many solutions.
Doing it eagerly makes it easy for the solver down the way but won't handle the case in which the index has a lot of solutions but only a few matter.

I most cases the lazy approach wont get too far neither if the index has a lot of solutions but it allows for the case in which you do not need to solve it (or in which the index gets constrained after the access?)

This is something the eagerly approach with fail at: 
```
index = input();
memory[10] = 100;
memory[200] = 200;
if (memory[index] < 150)
   printf( "Oh!");
else
  printf("Ah!");
```
